examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,149 @@
+        ac_f = ProbabilityOfImprovement()
+        botorch_acf = ac_f.to_botorch(
+            self.surrogate_model, searchspace, train_x, train_y
+        )
+        comp_rep_tensor = to_tensor(searchspace.discrete.comp_rep).unsqueeze(1)
+        poi = botorch_acf(comp_rep_tensor)
+        poi_list.append(poi)
+    else:
+        raise TypeError(
+            f"{searchspace.type} search spaces are not supported yet. "
+            f"Currently only DISCRETE search spaces are accepted."
+        )
+
+
+# Additionally, we define a function that plots the poi_list after all recommend  iterations:
+
+
+def plot_poi(poi_list: list):
+    """Plot the probability of improvement."""
+    num_plots = len(poi_list)
+    fig, axs = plt.subplots(num_plots, 1, figsize=(8, 6 * num_plots))
+
+    # Plot each POI tensor separately
+    for i, p in enumerate(poi_list):
+        poi_np = p.detach().numpy()
+        sns.histplot(
+            poi_np,
+            kde=True,
+            stat="density",
+            color="mediumaquamarine",
+            bins=20,
+            edgecolor="black",
+            linewidth=0.5,
+            ax=axs[i],
+        )
+        axs[i].set_xlabel(f"Iteration {i+1}: Probability of Improvement")
+        axs[i].set_ylabel("Search space Points")
+
+    fig.subplots_adjust(hspace=0.5)
+    plt.suptitle("Investigate the probability of improvement")
+    plt.show()
+
+
+### Monkeypatching
+
+# Next, we create our recommender and monkeypatch its `recommend` method:
+
+my_recommender = TwoPhaseMetaRecommender(
+    initial_recommender=RandomRecommender(),
+    recommender=BotorchRecommender(
+        surrogate_model=GaussianProcessSurrogate(),
+        allow_repeated_recommendations=True,
+        allow_recommending_already_measured=True,
+    ),
+)
+my_recommender.recommender.recommend = MethodType(
+    register_hooks(
+        BotorchRecommender.recommend,
+        post_hooks=[check_probability_of_improvement],
+    ),
+    my_recommender.recommender,
+)
+
+
+### Triggering the Hooks
+
+# We setup the other objects to trigger the hook:
+
+parameters = [
+    NumericalDiscreteParameter(name="x0", values=[1, 2, 3]),
+    NumericalDiscreteParameter(name="x1", values=[4, 5, 6]),
+    CategoricalParameter(name="Speed", values=["slow", "normal", "fast"]),
+]
+searchspace = SearchSpace.from_product(parameters=parameters)
+objective = SingleTargetObjective(target=NumericalTarget(name="yield", mode="MAX"))
+campaign = Campaign(
+    searchspace=searchspace,
+    recommender=my_recommender,
+    objective=objective,
+)
+
+# Now we run 5 recommend iterations to check the probability of improvement:
+
+for i in range(5):
+    recommendation = campaign.recommend(batch_size=3)
+    add_fake_results(recommendation, campaign)
+    campaign.add_measurements(recommendation)
+
+# Lastly, we plot the probability of improvement from the previous iterations to be able
+#  to analyse them:
+
+plot_poi(poi_list)
```

Comment:
When building the documentation, the plot is not shown, I think this is something that would be really nice.

Reply:
not only really nice but required, I'd say 🙃 I guess this requires some modification / extensions / decoupling of the plotting utility, @AVHopp?

Reply:
Why is it required? We do not show this in the older examples, so this cannot be that important (see e.g. `custom_analytical`):
![image](https://github.com/emdgroup/baybe/assets/84016733/c2046211-b9e3-4cf3-9d10-043399678021)
Still, including the plot one way or another (e.g. in the same way it is done in the "newer" examples) should be done here.
Regarding the decoupling: I agree that it would be nice to move that functionality to the corresponding file in the `utils` subpackage

Reply:
@RimRihana : The plot is still not being shown. I however do not see an immediate way of adressing this issue but I'll have a look such that you can focus on other things.

Reply:
I have found a way to solve this issue and pushed a corresponding commit: If the `poi_plot` function returns the corresponding `Figure` object that the `Axes` object belongs to, it is shown in the jupyter notebook. I do not know why this is not necessary for our other examples and particular for `create_examples_plot`, but well. Resolving this.